### PR TITLE
added test for testing binary data of children nodes

### PR DESCRIPTION
**Pull request by rederrik**
fails with error without https://github.com/jackalope/jackalope-doctrine-dbal/commit/851847b39cb363ca5cd0ea85c11301f6e622f931

### CODE REVIEW

File: tests/Writing/CopyMethodsTest.php [hunk 1]
```diff
@@ -239,4 +239,38 @@ public function testCopyUpdateReferencesMultiValue()
...
+        $srcChild = $this->session->getNode($src.'/data.bin');
+        $dstChild = $this->session->getNode($dst.'/data.bin');
+
+        $srcProp = $srcChild->getProperty('jcr:data');
+        $dstProp = $dstChild->getProperty('jcr:data');
+
+        $this->assertThat(true,
+            $this->logicalAnd(
+                $this->equalTo(\PHPCR\PropertyType::BINARY, $srcProp->getType()),
+                $this->equalTo(\PHPCR\PropertyType::BINARY, $dstProp->getType())
```
Comment (dbu) [inline tests/Writing/CopyMethodsTest.php]: why not just two $this->assertEquals?

File: tests/Writing/CopyMethodsTest.php [hunk 2]
```diff
@@ -239,4 +239,38 @@ public function testCopyUpdateReferencesMultiValue()
...
+            )
+        );
+
+        $srcBin = $srcProp->getBinary();
+        $dstBin = $dstProp->getBinary();
+
+        if (!is_resource($srcBin) || !is_resource($dstBin)) {
```
Comment (dbu) [inline tests/Writing/CopyMethodsTest.php]: $this->assertTrue(is_resource($srcBin), 'src is not a stream');
...

File: tests/Writing/CopyMethodsTest.php (hunk 1 above)
Comment (rederrik) [inline tests/Writing/CopyMethodsTest.php]: Instead of what?

Comment (dbu) [inline tests/Writing/CopyMethodsTest.php]: instead of assertThat(and(equal, equal)) you could do
```
$this->assertEquals(\PHPCR\PropertyType::BINARY, $srcProp->getType());
$this->assertEquals(\PHPCR\PropertyType::BINARY, $dstProp->getType());
```

your code seems correct, but more complicated than needed i think.

File: fixtures/10_Writing/copy.xml [hunk 3]
```diff
@@ -484,4 +484,23 @@
...
+        </sv:property>
+        <sv:node sv:name="srcNode">
+            <sv:property sv:name="jcr:primaryType" sv:type="Name">
+                <sv:value>nt:unstructured</sv:value>
+            </sv:property>
+            <sv:node sv:name="data.bin">
+                <sv:property sv:name="jcr:primaryType" sv:type="Name">
+                    <sv:value>nt:unstructured</sv:value>
+                </sv:property>
+                <sv:property sv:name="jcr:data" sv:type="Binary"><sv:value>RnVzY2UgZmVybWVudHVtLiBBbGlxdWFtIGxvYm9ydGlzLiBNYXVyaXMgdHVycGlzIG51bmMsIGJsYW5kaXQgZXQsIHZvbHV0cGF0IG1vbGVzdGllLCBwb3J0YSB1dCwgbGlndWxhLiBQZWxsZW50ZXNxdWUgYXVjdG9yIG5lcXVlIG5lYyB1cm5hLiBQcm9pbiBtYWduYS4gQ3VyYWJpdHVyIHVsbGFtY29ycGVyIHVsdHJpY2llcyBuaXNpLiBQcmFlc2VudCBjb25ndWUgZXJhdCBhdCBtYXNzYS4gTmFtIGFkaXBpc2NpbmcuIE51bGxhIHBvcnRhIGRvbG9yLiBGdXNjZSBjb21tb2RvIGFsaXF1YW0gYXJjdS4=
```
Comment (dbu) [inline fixtures/10_Writing/copy.xml]: i tried to run this with jackrabbit, but it fails with

exception 'Exception' with message 'java.lang.Exception: Failed to import repository to / from file /home/david/liip/jackalope/jackalope-jackrabbit/tests/../vendor/phpcr/phpcr-api-tests/fixtures/10_Writing/copy.xml
java.lang.IllegalArgumentException: specified data is not base64 encoded. 

the problem is the newline before `</sv:value>`.

when i change that in my code, things work.

File: tests/Writing/CopyMethodsTest.php (hunk 2 above)
Comment (rederrik) [inline tests/Writing/CopyMethodsTest.php]: I used $this->fail() to make sure the test wont go on.. not sure if I should use an assertion here.
Maybe just add a message?

File: fixtures/10_Writing/copy.xml (hunk 3 above)
Comment (rederrik) [inline fixtures/10_Writing/copy.xml]: Thanks, fixed!

File: tests/Writing/CopyMethodsTest.php (hunk 2 above)
Comment (dbu) [inline tests/Writing/CopyMethodsTest.php]: its the same. at the first assertion that fails, the test aborts.
semantically your code and what i propose would do the same, but i think
my version is easier to read, and you see which binary was not a resource

Comment (dbu) [inline tests/Writing/CopyMethodsTest.php]: btw, if you are unsure, just try
`$foo = null; $this->assertTrue(is_resource($foo)); stream_get_contents($foo); ` its always a good idea to play with your tests and make them fail to be sure you are doing what you meant to do :-)

Comment (rederrik) [inline tests/Writing/CopyMethodsTest.php]: Ah, you are right, thanks. 
Fixed!

**Comment by dbu** [comment]
thanks a lot! this now works in jackalope-jackrabbit, i squashed the commits and merged in 44405737dcecdf4e4760bb4472e4df56b49254de

**Comment by rederrik** [comment]
Thanks!

**Comment by lsmith77** [comment]
@dbu so we should close this PR?

**Comment by lsmith77** [comment]
did you also make a new release?

**Comment by dbu** [comment]
right, wanted to close this.

i did not release yet, we might need an additional test for the empty array multivalue binary case.
